src/conv/heuristics/metadata_3d.cpp

```diff
@@ -0,0 +1,348 @@
+    const_cast<std::unordered_map<std::string, int>&>(in_layout_encodings) = std::move(*in_layout_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(fil_layout_encodings) = std::move(*fil_layout_encodings_opt);
+    const_cast<std::unordered_map<std::string, int>&>(out_layout_encodings) = std::move(*out_layout_encodings_opt);
+    
+    // Mark as valid after successful loading
+    is_valid = true;
+    
+    if (miopen::IsLogging(LoggingLevel::Info2)) {
+        MIOPEN_LOG_I2("Metadata3D loaded successfully for arch: " << arch << 
+                     ", num_inputs=" << num_inputs << ", num_solvers=" << num_solvers);
+    }
+}
+
+// Encoding methods with safe error handling
+size_t Metadata3D::EncodeDirection(miopen::conv::Direction dir) const
+{
+    if (!is_valid) return 0;
+    
+    try {
+        if(dir == conv::Direction::BackwardWeights)
+            return direction_encodings_3d.at("W");
+        else if(dir == conv::Direction::BackwardData)
+            return direction_encodings_3d.at("B");
+        else
+            return direction_encodings_3d.at("F");
+    } catch (...) {
+        MIOPEN_LOG_W("Direction encoding failed in 3D metadata, returning 0");
+        return 0;
+    }
+}
+
+size_t Metadata3D::EncodePrecision(miopenDataType_t data_type) const
+{
+    if (!is_valid) return 0;
+    
+    try {
+        if(data_type == miopenBFloat16)
+            return precision_encodings_3d.at("BF16");
+        else if(data_type == miopenHalf)
+            return precision_encodings_3d.at("FP16");
+        else if(data_type == miopenFloat)
+            return precision_encodings_3d.at("FP32");
+        else {
+            MIOPEN_LOG_W("Unsupported data type in 3D metadata, returning 0");
+            return 0;
+        }
+    } catch (...) {
+        MIOPEN_LOG_W("Precision encoding failed in 3D metadata, returning 0");
+        return 0;
+    }
+}
+
+size_t Metadata3D::EncodeLayout(const std::string& layout) const
+{
+    if (!is_valid) return 0;
+    
+    auto it = in_layout_encodings.find(layout);
+    if(it != in_layout_encodings.end())
+        return it->second;
+    
+    MIOPEN_LOG_W("Unsupported layout " << layout << " in 3D metadata, returning 0");
+    return 0;
+}
+
+size_t Metadata3D::EncodeInLayout(const std::string& layout) const
+{
+    if (!is_valid) return 0;
+    
+    auto it = in_layout_encodings.find(layout);
+    return (it != in_layout_encodings.end()) ? it->second : 0;
+}
+
+size_t Metadata3D::EncodeFilLayout(const std::string& layout) const
+{
+    if (!is_valid) return 0;
+    
+    auto it = fil_layout_encodings.find(layout);
+    return (it != fil_layout_encodings.end()) ? it->second : 0;
+}
+
+size_t Metadata3D::EncodeOutLayout(const std::string& layout) const
+{
+    if (!is_valid) return 0;
+    
+    auto it = out_layout_encodings.find(layout);
+    return (it != out_layout_encodings.end()) ? it->second : 0;
+}
+
+} // namespace conv3d
+} // namespace ai
+} // namespace miopen
```

Comment:
```suggestion
} // namespace miopen

```
the formatter prefers a newline at the end of the file